### PR TITLE
ISSUE-120: Added an admin hint block and tab for the ADO view mode display

### PR DESCRIPTION
**Pull request by pcambra**
This PR adds the following:

- New block `DisplayInformationBlock` that shows the view mode selected linked to the display settings of that view mode
- A local task plugin to inject the right parameters to the Display settings tab per ADO (`ViewModeLocalTask`)
- A consolidated service: `ViewModeResolver` that is called in view_mode_alter but also in the informative block, it does the heavy lifting of selecting the view mode, caching the biggest and common part (getADOTypes)

### CODE REVIEW

**Comment by pcambra** [comment]
Resolves #120

**Comment by DiegoPino** [comment]
@pcambra I will review this during the weekend, thanks a lot!
We will discuss next week some performance concerns I have with this new implementation, probably better with numbers in place but may be just small things. Have a great weekend

**Comment by pcambra** [comment]
@DiegoPino this is ready to review again :)

File: format_strawberryfield.module
Comment (DiegoPino) [inline format_strawberryfield.module:47]: good. Can't remember why I left that there. Thanks for removing

File: format_strawberryfield.routing.yml
Comment (DiegoPino) [inline format_strawberryfield.routing.yml:235]: Tiny thing. These are Display settings for bundles of type ADO, not for each ADO right?

File: src/Plugin/Block/DisplayInformationBlock.php
Comment (DiegoPino) [inline src/Plugin/Block/DisplayInformationBlock.php:113]: Let's change this to "Info"

File: src/Plugin/Block/DisplayInformationBlock.php
Comment (DiegoPino) [inline src/Plugin/Block/DisplayInformationBlock.php:139]: lovely, all that uses caches is lovely

File: src/Plugin/Menu/LocalTask/ViewModeLocalTask.php
Comment (DiegoPino) [inline src/Plugin/Menu/LocalTask/ViewModeLocalTask.php:56]: nice!

File: src/ViewModeResolver.php
Comment (DiegoPino) [inline src/ViewModeResolver.php:18]: let's add a an extra line here before the trait , just for readability (sorry, I'm lacking coffee) 👴

File: src/ViewModeResolver.php
Comment (DiegoPino) [inline src/ViewModeResolver.php:74]: Is default the right one? Where are the saved "view modes" stored when a user selects instead of the default (during edit/create) one?

File: src/ViewModeResolver.php
Comment (DiegoPino) [inline src/ViewModeResolver.php:89]: This cache should be per "JSON type" not per entity ID: on a 400,000 entity site the cache instead of helping will become a burden. If ADO or not is at the end a bundle type definition, and which view mode a JSON key "type" value one.

**Review by DiegoPino — CHANGES_REQUESTED** [review]
Hi @pcambra thanks. I added mostly comments except for a larger one which requires intervention/code change (thus a request changes state here).

The cache can not be by entity->id(() If you go that route that cache will outlive its usefulness at generate a huge entry. The cache should be by "bundle identified as ADO" and JSON type found inside the flattened structure of the JSON. That way your cache will be Max of size of the different Json types x the view modes assigned x bundles bearing a SBF.

File: src/ViewModeResolver.php
Comment (pcambra) [reply src/ViewModeResolver.php:74]: I'm afraid that comes from display suite, `$node->ds_switch->value`, and I'm not sure we want that dependency. Something like `\Drupal::service('entity_display.repository')->getViewDisplay($entity_type, $bundle, $view_mode);` will get us the default bundle display, but I'm not sure if it's enough.

File: src/ViewModeResolver.php
Comment (pcambra) [reply src/ViewModeResolver.php:89]: @DiegoPino I don't see how this can be cached this way, as we're dependent on the entity. On this list we're caching the ADO types per entity as `bearsStrawberryfield` service still gets it as parameter. Seems like a catch 22, I can't cache before calling and I don't know the necessary information before caching.

Comment (DiegoPino) [reply src/ViewModeResolver.php:89]: The egg/chicken dilemma (catch 22) is the reason we did not apply a cache previously for the json type itself when just having an entity. (the lack of a proper class of course at bundle level). Let me think about this a little deeper (like 24 hours, busy day preparing for tomorrow many workshop/zoom type of things) and give the whole thing a spin. I'm pretty sure there may be a pattern that is escaping us where performance and atomicity can play well. Thanks

**Comment by pcambra** [comment]
@DiegoPino as an aside conversation, I've found that we're using the same code in https://github.com/esmero/ami/blob/main/src/AmiUtilityService.php#L253, we might want to do something about it

File: src/ViewModeResolver.php
Comment (pcambra) [reply src/ViewModeResolver.php:89]: I've been thinking about this for a bit, as if the solution of caching it per entity is not acceptable, having a computed field wouldn't be either, I think that giving how aggressively Drupal caches everything, this would be fine but you obviously know the system better. Drupal performs so so with a lot of content without external caching and for memcache/redis, 400k simple entries like these are a breeze.

In any case, this is providing an additional cache layer that we don't currently have, so I would suggest three things here:

- Removing the cache layer.
- Adding a followup issue where we consider the field approach instead (this would replace the service)
- Having a place where we can actually test this; maybe an online version of a mid sized repository with something like new relic or the symfony one, where we can benchmark stuff?

**Comment by DiegoPino** [comment]
@pcambra AMI has a lot of legacy code. That method is not in use right now and I have myself put the task of cleaning it up during this week.

That function is actually 1:1 to the view mode except its not about the view mode but about the actual type based on an UUID (AMI uses UUIDs to reference entities because IDs match also spreadsheet row numbers and it would be a mess, but in general UUIDs are better). The idea of that method (again, not used) is to fetch the JSON type key of a related (ismemberof, parof) Node and use semantics (not coded) to allow/now allow such relationships. But even if I have a full complex way of doing that via ontologies its really out of the scope of our work during 1.0.0. Ontologies (like "a bird can have wings and only 2 but not all things that have wings are birds") are cool and nice but (experience) will go over the head of most of our users, still working their way around JSON and Twig templates. Sorry long explanation and thanks for detecting that one.

File: src/ViewModeResolver.php
Comment (DiegoPino) [reply src/ViewModeResolver.php:89]: > - Removing the cache layer.
Ok. I agree. Let's remove the cache layer for now. Caching which bundles are of ADO type may something easier we can approach (at least one check less) and that can go into the actual `::bearsStrawberryfield` method (another issue, not to be solved here)
> - Adding a followup issue where we consider the field approach instead (this would replace the service)

The field approach you mean for saving the view mode at the Node level? Instead of using the ds-> config? Yes. may even an option that people want to opt/in/out. We may keep the service and build a field check there if we go that route. Better to have more options than less.

> - Having a place where we can actually test this; maybe an online version of a mid sized repository with something like new relic or the symfony one, where we can benchmark stuff?
I agree. We can feed one of our repos a lot of data and even create a test batch ingest set (via the JSON API, a special costume module) so we can benchmark this. I once created 1,000,000 random objects in 2018 for this (when I decided on this architecture) but it was a brute force approach. Should be see this as a DevOps task with @alliomeria collaboration for the data?

File: src/ViewModeResolver.php
Comment (DiegoPino) [reply src/ViewModeResolver.php:18]: Yeah, I had my coffee(s). Not even worth the change. Dismiss!

File: src/Plugin/Block/DisplayInformationBlock.php
Comment (DiegoPino) [reply src/Plugin/Block/DisplayInformationBlock.php:113]: @pcambra just a few things pending here before I can merge. The Warning to Info is important so we do not scare the users!

File: src/ViewModeResolver.php
Comment (DiegoPino) [reply src/ViewModeResolver.php:89]: I will mark the cache discussion as resolved for now until we have larger numbers and we can measure if cache v/s dynamically calculating this every time changes or not things. Thanks a lot of https://github.com/esmero/format_strawberryfield/pull/122/commits/bc84fe57d9fea3ca18f73f719159b824254bfc4d Which was needed.

**Review by DiegoPino — APPROVED** [review]
@pcambra approving this, but won't merge until the "Warning" v/s "info" message is changed. Thanks a lot, I know this was hard work and I appreciate the level of detail in your code!

**Comment by pcambra** [comment]
@DiegoPino apologies, I'd swear I had changed that

**Comment by pcambra** [comment]
@DiegoPino this should be ready to go I think

**Pull request merged**
